### PR TITLE
build redirection file with versioning info

### PR DESCRIPTION
**Pull request by 928PJY**
this fix is to handle the following two cases:
1. there are two b.md with different versioning: v1/b.md and v2/b.md
and there are two redirection(with_document_id: true) rules:
```yml
v1/a.md -> /b
v2/a.md -> /b
```
in current docfx v3, it will be treated as redirection-url-conflict.

2. there are two b.md with different versioning: v1/b.md and v2/b.md
and there are one redirection(with_document_id: true) rules:
```yml
v1/a.md -> /b
```
in current docfx v3, the document_id of v1/a.md will be taken to v1/b.md(expected) and **v2/b.md**(unexpected)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5126)

### CODE REVIEW

**Comment by yufeih** [comment]
This PR and https://github.com/dotnet/docfx/pull/5136 makes me think that we might want to introduce a _site url identifier_ that takes versioning into account.

File: src/docfx/build/redirection/RedirectionMap.cs [hunk 1]
```diff
@@ -126,27 +125,48 @@ private static string NormalizeRedirectUrl(string redirectionUrl)
...
+                .GroupBy(file => file.SiteUrl)
+                .ToDictionary(group => group.Key, group => group.ToList(), PathUtility.PathComparer);
 
-            var publishUrls = buildFiles.Select(file => file.SiteUrl)
-                .Concat(redirections.Select(item => item.SiteUrl)).ToHashSet();
-            foreach (var (originalRedirectUrl, normalizedRedirectUrl) in redirectionsWithDocumentId)
+            foreach (var (originalRedirectUrl, redirect) in redirectionsWithDocumentId)
             {
-                if (!publishUrls.Contains(normalizedRedirectUrl))
+                var (_, redirectionSourceMonikers) = monikerProvider.GetFileLevelMonikers(redirect);
```
[the remark this answers is not in the document]
Comment (live1206) [reply src/docfx/build/redirection/RedirectionMap.cs]: OK. And we should never discard errors from the lower level functions?

[the remark this answers is not in the document]
Comment (yufeih) [reply src/docfx/build/redirection/RedirectionMap.cs]: yep

**Comment by 928PJY** [comment]
there is some difference between the `publish-url-conflict` and `redirection-url-conflict`
the `publish-url-conflict` is monikers of files with the same url cannot be overlapping.
the `redirection-url-conflict` is the document indicated by the redirection_url cannot be same.
and seems to bring in the concept `site url identifier` cannot consociate those two cases.

---
In reply to: [534830510](https://github.com/dotnet/docfx/pull/5126#issuecomment-534830510) [](ancestors = 534830510)

**Pull request merged**